### DISCUSSION TITLE
tinyMCE custom icon alternate state

### DISCUSSION THREAD

**Discussion by Philo-U** [tinymce/tinymce#10001] [General]
Hi - I asked this on StackOverflow, but no resposne. so hopefully someone here can help...

So I can add a custom toolbar toggle button that will render the selected text in red - viz:
```
        tinymce.init({
              selector: '#txtT',
              width: 500,
              height: 300,
              browser_spellcheck: true,
              statusbar: false,
              menubar: false,
              toolbar: 'bold italic underline | boldRed',
              setup: (editor) => {
                  editor.ui.registry.addIcon('mysvg', '<svg height="24" width="24"><path d="M12 0 L24 24 L0 24 Z" /></svg>');
                  editor.ui.registry.addToggleButton('boldRed', {
                      icon: 'mysvg',
                      onAction: (_) => editor.execCommand('ForeColor', false, '#FF0000'),
                      onSetup: (api) => {
                          api.setActive(editor.formatter.match('#FF0000'));
                          const changed = editor.formatter.formatChanged('#FF0000', (state) => api.setActive(state));
                          return () => changed.unbind();
                      }
                  });
              },
              content_style: 'body { font-family:Helvetica,Arial,sans-serif; font-size:16px }'
          });
```
So far so good. Two things though:

1. the default buttons, such as B (bold) will switch to an alternate, "on", version if the cursor is placed over bold text, giving a visual clue to the format of the text at the cursor point. 
I cannot figure out how to get my custom icon to do this. Presumably I need to register the alternate, toggled, "on" state icon somehow ..?

Even if I use an in-built icon - say "bold" (again) - this doesn't change state the way the real bold, italic etc ones do. S I guess  this too needs referencing in my code somewhere...?

Pretty sure the answer lies with [context](https://www.tiny.cloud/docs/tinymce/latest/context/), but damned if I can figure it out...

2. ideally, I really want this button to render the text in bold red, not just bold (and revert to normal black when toggled)... am a bit lost on this too...?

Added a [fiddle here](https://jsfiddle.net/578dp2ar/1/)

Any help gratefully received....

**Comment by lostkeys** [comment]
Hey @Philo-U.

To have the button become enabled the way you want, you need to defina a proper format. Here is an example: https://fiddle.tiny.cloud/jAqKcisWxj/0

**Reply by Philo-U** [reply]
Thank you!!
